### ISSUE TITLE
setdefault on TypedDicts

### ISSUE THREAD

**Issue by JelleZijlstra** [microsoft/pyright#2875]
```python
from typing_extensions import TypedDict

class TD(TypedDict):
    a: int
    b: str

def typeddict(td: TD):
    td.setdefault("c")
    td.setdefault("a", "s")
```
Pyright should produce an error on both `setdefault` lines, but it doesn't. Setting `total=False` doesn't make a difference.

Unfortunately pretty much every dict method needs special casing on TypedDict.

**Comment by erictraut** [comment]
Pyright auto-generates overloads for `setdefault` based on the TypedDict entries. However, it adds a fallback overload to handle dynamic key values (as opposed to literal values). That fallback overload is what's allowing these to pass without an error. I could delete the fallback overload, but that would eliminate the possibility of handling dynamic keys. It looks like mypy flags the use of a dynamic key as an error, so maybe that's acceptable.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by JelleZijlstra** [comment]
Thanks! https://www.python.org/dev/peps/pep-0589/#supported-and-unsupported-operations does say that "A key that is not a literal should generally be rejected".

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.210, which I just published. It will also be included in the next release of pylance.

**Comment by GBeauregard** [comment]
The `get` bugfix committed with this is a bit annoying in strict mode, since you can't do `myvar = myjson.get("library_versions", [])` in strict mode anymore since the `[]` needs typed, but typing `[]` inline is a bit ugly and means I'm repeating the info I took care to place into the TypedDict: `myvar = myjson.get("library_versions", list[Mapping[str,int]]())`. 

I realize this is probably intended, but this situation is unsatisfying. Can there be a heuristic for empty literals compatible with the typeddict key's type? Some other solution?

**Comment by erictraut** [comment]
How about `myjson.get("library_versions", None) or []`?

**Comment by erictraut** [comment]
An alternative is to undo that part of this change and disallow specifying a default value whose type differs from the declared type for that field, but that seems to differ from the intent of the function (and the behavior provided by mypy).

**Comment by GBeauregard** [comment]
> How about `myjson.get("library_versions", None) or []`?

Hmmm...maybe, but I don't particularly like it and I still end up having to retype the LHS

I worked on a minimal example to tease out the specific annoyances.

One is the obvious one I mentioned before:

```python
# pyright: strict
from typing import TypedDict

class MyJSON(TypedDict):
    streams: list[str]

def returns_json() -> MyJSON: ...

a = returns_json().get("streams", [])  # Argument type is partially unknown
```

> An alternative is to undo that part of this change and disallow specifying a default value whose type differs from the declared type for that field, but that seems to differ from the intent of the function (and the behavior provided by mypy).

I agree with you on the intent, but I find it hard to imagine this justifies the annoyance from usage in practice. However, I don't want to break people with weird default values needlessly. I think a heuristic where if the `[]` is compatible with the declared type then it works like before would be my preferred solution

Another aspect of this that came up while investigating is `[]` interacts oddly (but not 'wrong' per se) with a TypedDict marked Sequence:

```python
# pyright: strict
from collections.abc import Sequence
from typing import TypedDict

class MyJSON(TypedDict):
    streams: Sequence[str]

def returns_json() -> MyJSON:
    ...

a = returns_json().get("streams", [])
```

**Comment by erictraut** [comment]
I'm unwilling to add such a strange and specific heuristic in this case. The evaluation of `[]` is in a part of the code that has absolutely no relation to TypedDict's, so consider that option off the table.

The problem that you're facing here isn't specific to `TypedDict.get`. It is related to many uses of the expression `[]` or `{}`. I do have some special-casing in place for these already to handle the following:

```python
# pyright: strict
def func(cond: bool):
    if cond:
        x = []
    else:
        x = [1, 2, 3]
    reveal_type(x) # list[int]
```

I could generally suppress the "partially unknown" diagnostic if `[]` or `{}` are passed as arguments to a function that accepts a concrete `list` or `dict` type, but that wouldn't work in this case because the receiving parameter's type is a TypeVar.

I think this particular case that you're hitting is pretty specific to your use case, so unless you have any better suggestions, I think you'll need to just deal with the impact of the recent bug fix. Sorry about that.

**Comment by GBeauregard** [comment]
Yes, the `or` workaround isn't the worst thing in the world (although it would be better with the None-aware operators), but I will continue thinking about this while keeping in mind the relevant code isn't special-cased.

**Comment by erictraut** [comment]
The special-casing logic for `[]` and `{}` is why `a` is resolved as `list[str]` rather than `list[str] | list[unknown]` in your example above. This same special case doesn't apply to `Sequence`, which explains why you're seeing the expected result of `Sequence[str] | list[unknown]` in that case.

Thinking about this a little more, I think it's safe to always suppress the "partially unknown argument type" diagnostic if I have a `list[unknown]` or `dict[unknown]` that came from a `[]` or `{}` expression. This would eliminate the strict error in at least some of your cases. It still wouldn't help with the `Sequence` example though.

**Comment by GBeauregard** [comment]
That suppression is a good start that probably handles the most common stuff.

--- 
Ahah, yes, that fully resolves my confusion on how the two cases ended up different. But, the `or` workaround seems to be unexpectedly effective in the Sequence case:

```python
# pyright: strict
from collections.abc import Sequence
from typing import TypedDict

class MyJSON(TypedDict):
    streams: Sequence[str]

def returns_json() -> MyJSON: ...

a = returns_json().get("streams", None) or []
reveal_type(a)  # Sequence[str] | list[str]
```

expected (based on special case not applying): `reveal_type(a)  # Sequence[str] | list[unknown]`, so maybe my understanding isn't as good as I thought

**Comment by erictraut** [comment]
The `or` case works because I apply bidirectional type inference for the right operand's expression based on the type of the left operand for binary operations (including `or` and `and`). For example:

```python
x: list[str | int] = ["hi", 1]
y = x or [1]
reveal_type(y) # `list[str | int]`, not `list[str | int] | list[int]`
```

Likewise:
```python
x: Sequence[int] = []
y = x or []
reveal_type(y) # `Sequence[int] | list[int]`, not `Sequence[int] | list[Unknown]`
```
